### PR TITLE
Backport TUF security bugfix to 3.6

### PR DESCRIPTION
**Pull request by andreabedini**
cherry-picks dcfdc9cffd74cade4e8cf3df37c5993413ffd30f

closes: #9451

### CODE REVIEW

**Review by Kleidukos — APPROVED** [review]

File: cabal-install/src/Distribution/Client/CmdUpdate.hs [hunk 1]
```diff
@@ -197,10 +197,12 @@ updateRepo verbosity _updateFlags repoCtxt (repo, indexState) = do
       -- NB: always update the timestamp, even if we didn't actually
       -- download anything
       writeIndexTimestamp index indexState
-      ce <- if repoContextIgnoreExpiry repoCtxt
-              then Just `fmap` getCurrentTime
-              else return Nothing
-      updated <- Sec.uncheckClientErrors $ Sec.checkForUpdates repoSecure ce
+      -- typically we get the current time to check expiry against
+      -- but if the flag is set, we don't.
+      now <- case repoContextIgnoreExpiry repoCtxt of
```
Comment (hasufell) [inline cabal-install/src/Distribution/Client/CmdUpdate.hs:202]: This actually causes name-shadowing just like in the 3.10 commit: https://github.com/haskell/cabal/commit/dcfdc9cffd74cade4e8cf3df37c5993413ffd30f

See line 213.

Comment (andreabedini) [reply cabal-install/src/Distribution/Client/CmdUpdate.hs:202]: Sorry. I did this late at night. I'll push a fix but I won't be able to do it today.
I won't be offended if someone else did it before me ☺️

File: cabal-install/src/Distribution/Client/CmdUpdate.hs [hunk 2]
```diff
@@ -197,10 +197,12 @@ updateRepo verbosity _updateFlags repoCtxt (repo, indexState) = do
...
+                 False -> Just <$> getCurrentTime
+                 True  -> pure Nothing
+      updated <- Sec.uncheckClientErrors $ Sec.checkForUpdates repoSecure now
```
Comment (hasufell) [inline cabal-install/src/Distribution/Client/CmdUpdate.hs:205]:
```suggestion
      updated <- do
        now <- case repoContextIgnoreExpiry repoCtxt of
                   False -> Just <$> getCurrentTime
                   True  -> pure Nothing
        Sec.uncheckClientErrors $ Sec.checkForUpdates repoSecure now
```

Comment (hasufell) [reply cabal-install/src/Distribution/Client/CmdUpdate.hs:205]: Ah, it seems it was later fixed differently: https://github.com/haskell/cabal/commit/ec86ddcdd990b358ada9395c41a94a408c5eb9a8

No idea why the commit messages are unhelpful.

Comment (andreabedini) [reply cabal-install/src/Distribution/Client/CmdUpdate.hs]: Yeah, commit messages and PRs left me in the dark. I used the same version as in https://github.com/haskell/cabal/commit/ec86ddcdd990b358ada9395c41a94a408c5eb9a8 just to keep things consistent. Thank you.

**Comment by andreabedini** [comment]
I had to restrict the bound of hackage-security to get this to work. 

hackage-security-0.6.2.0 added a dependency on Cabal-syntax, which is incompatible with Cabal-3.6.

**Comment by hasufell** [comment]
> I had to restrict the bound of hackage-security to get this to work.
> 
> hackage-security-0.6.2.0 added a dependency on Cabal-syntax, which is incompatible with Cabal-3.6.

It's not, you can use this patch:

```diff
From 9a04c42a76883f141bc4589a2d26b28c63cde10c Mon Sep 17 00:00:00 2001
From: Julian Ospald <hasufell@posteo.de>
Date: Thu, 16 Nov 2023 21:00:26 +0800
Subject: [PATCH 1/5] Fix build with newer GHC

---
 cabal-install/src/Distribution/Client/Win32SelfUpgrade.hs | 2 +-
 cabal.project.release                                     | 3 +++
 cabal.project.validate                                    | 2 ++
 3 files changed, 6 insertions(+), 1 deletion(-)

diff --git a/cabal-install/src/Distribution/Client/Win32SelfUpgrade.hs b/cabal-install/src/Distribution/Client/Win32SelfUpgrade.hs
index 0d9a9effe..c220d9e92 100644
--- a/cabal-install/src/Distribution/Client/Win32SelfUpgrade.hs
+++ b/cabal-install/src/Distribution/Client/Win32SelfUpgrade.hs
@@ -140,7 +140,7 @@ deleteOldExeFile verbosity oldPID tmpPath = do
      ++ show oldPID ++ " at path " ++ tmpPath
 
   log $ "getting handle of parent process " ++ show oldPID
-  oldPHANDLE <- Win32.openProcess Win32.sYNCHORNIZE False (fromIntegral oldPID)
+  oldPHANDLE <- Win32.openProcess Win32.sYNCHRONIZE False (fromIntegral oldPID)
 
   log $ "synchronising with parent"
   event <- openEvent syncEventName
diff --git a/cabal.project.release b/cabal.project.release
index 4fda782ae..8a5f856a3 100644
--- a/cabal.project.release
+++ b/cabal.project.release
@@ -3,3 +3,6 @@ packages: cabal-install/
 tests: False
 benchmarks: False
 optimization: True
+constraints: Cabal-syntax < 3.7,
+             Win32 >= 2.8
+allow-newer: base
diff --git a/cabal.project.validate b/cabal.project.validate
index 25f29243d..33b28876f 100644
--- a/cabal.project.validate
+++ b/cabal.project.validate
@@ -17,6 +17,8 @@ tests: True
 -- avoiding extra dependencies
 constraints: rere -rere-cfg
 constraints: these -assoc
+constraints: Cabal-syntax < 3.7,
+             Win32 >= 2.8
 
 write-ghc-environment-files: never
 
-- 
2.41.0
```

**Comment by andreabedini** [comment]
@hasufell 
> It's not, you can use this patch:

Sorry you need to be more explicit with me. Do you mean you want to allow cabal-install-3.6 to build against a newer Cabal than 3.6? I am not opposed to that but I think it can be discussed separately.

I don't understand what you are suggesting, the patch above changes a call to Win32.sYNCHRONIZE, how is this related to hackage-security?

If I am not mistaken cabal-install-3.6.2.0 has been unbuildable since hackage-security-0.6.2.0 came out so that is an urgent concern (we should do a revision on hackage too).

**Comment by hasufell** [comment]
> Sorry you need to be more explicit with me. 

This fixes it:

```
constraints: Cabal-syntax < 3.7
```

Cabal-syntax-3.6 is a dummy package:

> Version 3.6 (unlike the following versions) is a dummy package that prevents module name clases between Cabal and Cabal-syntax if used together with a Cabal flag as described below.

The Win32 thing was needed to fix builds on windows for me.

**Comment by andreabedini** [comment]
@hasufell that a project configuration setting, it is not part of the package metadata.

**Comment by hasufell** [comment]
> @hasufell that a project configuration setting, it is not part of the package metadata.

You can put it in the package metadata.

File: cabal-install/cabal-install.cabal [hunk 3]
```diff
@@ -290,7 +290,7 @@ executable cabal
         time       >= 1.5.0.1  && < 1.12,
         transformers >= 0.4.2.0 && < 0.6,
         zlib       >= 0.5.3    && < 0.7,
-        hackage-security >= 0.6.0.1 && < 0.7,
+        hackage-security >= 0.6.0.1 && < 0.6.2.0,
```
Comment (hasufell) [inline cabal-install/cabal-install.cabal:293]: It builds fine with 0.6.2.3.

Comment (andreabedini) [reply cabal-install/cabal-install.cabal:293]: It does not:
```
$ cd $(mktemp -d) && echo -n 'extra-packages: cabal-install-3.6.2.0' > cabal.project && ghcup run --ghc 9.2.8 -- cabal build cabal-install --index-state=2023-11-17T02:00:58Z --constraint hackage-security==0.6.2.3
Warning: There are no packages or optional-packages in the project
Resolving dependencies...
Build profile: -w ghc-9.2.8 -O1
In order, the following will be built (use -v for more details):
 - cabal-install-3.6.2.0 (exe:cabal) (requires build)
Starting     cabal-install-3.6.2.0 (exe:cabal)
Building     cabal-install-3.6.2.0 (exe:cabal)

Failed to build exe:cabal from cabal-install-3.6.2.0.
...
```

Comment (andreabedini) [reply cabal-install/cabal-install.cabal:293]: Thank you @andreasabel

**Comment by andreabedini** [comment]
I tried to get the CI to pass but I have given up :)

**Comment by ulysses4ever** [comment]
What's the plan for this? I'm not sure it's the best application of our scattered resources to patch the old releases.

**Comment by Kleidukos** [comment]
@andreabedini Let's leave 3.6 to be patched by downstream if they wish so, it would be better to focus on delivering 3.10.3.0

**Comment by andreabedini** [comment]
> @andreabedini Let's leave 3.6 to be patched by downstream if they wish so, it would be better to focus on delivering 3.10.3.0

:+1:

**Pull request closed without merging**